crates/solver/src/settlement_rater.rs

```diff
@@ -244,18 +248,251 @@ impl SettlementRating for SettlementRater {
+            }
+        };
+
         let rated_settlement = RatedSettlement {
             id,
             settlement,
             surplus: inputs.surplus_given,
             earned_fees,
             solver_fees: inputs.solver_fees,
             gas_estimate: simulation.gas_estimate,
-            gas_price: effective_gas_price.clone(),
+            gas_price: effective_gas_price,
+            gas_cost,
             objective_value,
             score,
             ranking: Default::default(),
         };
         Ok(Rating::Ok(rated_settlement))
     }
 }
+
+pub struct ScoreCalculator {
+    score_cap: BigRational,
+    submission_strategies: Vec<TransactionStrategyArg>,
+}
+
+impl ScoreCalculator {
+    pub fn new(score_cap: BigRational, submission_strategies: Vec<TransactionStrategyArg>) -> Self {
+        Self {
+            score_cap,
+            submission_strategies,
+        }
+    }
+
+    pub fn cost_fail(&self, gas_cost: &BigRational) -> BigRational {
+        if self.submission_strategies == vec![TransactionStrategyArg::Flashbots] {
+            BigRational::zero()
+        } else {
+            gas_cost.clone()
+        }
+    }
+
+    pub fn compute_score(
+        &self,
+        objective_value: &BigRational,
+        gas_cost: &BigRational,
+        success_probability: f64,
+    ) -> Result<Score> {
+        ensure!(
+            (0.0..=1.0).contains(&success_probability),
+            "success probability must be between 0 and 1."
+        );
+        let success_probability = BigRational::from_float(success_probability).unwrap();
+        let cost_fail = self.cost_fail(gas_cost);
+        let optimal_score =
+            self.compute_optimal_score(objective_value.clone(), success_probability, cost_fail)?;
+        let score = big_rational_to_u256(&optimal_score).context("Invalid score.")?;
+        Ok(Score::ProtocolWithRisk(score))
+    }
+
+    fn compute_optimal_score(
+        &self,
+        objective: BigRational,
+        probability_success: BigRational,
+        cost_fail: BigRational,
+    ) -> Result<BigRational> {
+        tracing::trace!(
+            ?objective,
+            ?probability_success,
+            ?cost_fail,
+            "Computing optimal score"
+        );
+
+        let profit = |score_reference: BigRational| {
+            self.profit(
+                score_reference,
+                objective.clone(),
+                probability_success.clone(),
+                cost_fail.clone(),
+            )
+        };
+
+        let profit_cap = profit(self.score_cap.clone());
+        let profit_obj_minus_cap = profit(objective.clone() - self.score_cap.clone());
+        tracing::trace!(
+            ?profit_obj_minus_cap,
+            ?profit_cap,
+            "profit score minus cap and profit cap"
+        );
+
+        let probability_fail = BigRational::one() - probability_success.clone();
+
+        // https://www.notion.so/cownation/Optimal-bidding-strategy-84b4c710466a4c56af9295015308452a
```

Comment:
Can we publish this or better even check it in to the code as an MD file so that not only core contributors can read this?